app/scripts/controller-init/utils.ts

```diff
@@ -0,0 +1,169 @@
+import { createProjectLogger } from '@metamask/utils';
+import {
+  BaseControllerMessenger,
+  BaseRestrictedControllerMessenger,
+  ControllerByName,
+  ControllerInitFunction,
+  ControllerInitRequest,
+  ControllerName,
+} from './types';
+import { Controller } from './controller-list';
+import { CONTROLLER_MESSENGERS } from './messengers';
+
+const log = createProjectLogger('controller-init');
+
+/** Result of initializing controllers. */
+export type InitControllersResult = {
+  /** All API methods exposed by the controllers. */
+  controllerApi: Record<string, Controller>;
+
+  /** All controllers that provided a memory state key. */
+  controllerMemState: Record<string, Controller>;
+
+  /** All controllers that provided a persisted state key. */
+  controllerPersistedState: Record<string, Controller>;
+
+  /** All initialized controllers keyed by name. */
+  controllersByName: ControllerByName;
+};
+
+type BaseControllerInitRequest = ControllerInitRequest<
+  BaseRestrictedControllerMessenger,
+  BaseRestrictedControllerMessenger
+>;
+
+type BaseControllerInitFunction = ControllerInitFunction<
+  Controller,
+  BaseRestrictedControllerMessenger,
+  BaseRestrictedControllerMessenger
+>;
+
+type ControllerMessengerCallback = (
+  BaseControllerMessenger: BaseControllerMessenger,
+) => BaseRestrictedControllerMessenger;
+
+type ControllersToInitialize = 'PPOMController' | 'TransactionController';
+
+/**
+ * Initialize the controllers according to the provided init objects.
+ * Each init object can be a function that returns a controller, or a `ControllerInit` instance.
+ *
+ * @param options - Options bag.
+ * @param options.baseControllerMessenger - Unrestricted base controller messenger.
+ * @param options.existingControllers - All required controllers that have already been initialized.
+ * @param options.initFunctions - Map of init functions keyed by controller name.
+ * @param options.initRequest - Base request used to initialize the controllers.
+ * Excluding the properties that are generated by this function.
+ * @returns The initialized controllers and associated data.
+ */
+export function initControllers({
+  baseControllerMessenger,
+  existingControllers = [],
+  initFunctions,
+  initRequest,
+}: {
+  baseControllerMessenger: BaseControllerMessenger;
```

Comment:
It'd be nice to use a more-correct type here as well, but I don't think we should bother attempting this until the `Controllers` type includes everything. It wouldn't be easy to make this correctly typed before then.